File: cwt-spring/src/main/java/com/cwtsite/cwt/message/entity/Message.java
package com.cwtsite.cwt.message.entity;

import com.cwtsite.cwt.message.entity.enumeration.MessageCategory;
import com.cwtsite.cwt.user.repository.entity.User;
import org.hibernate.annotations.CreationTimestamp;

import javax.persistence.*;
import java.sql.Timestamp;
import java.util.List;

@Entity
@Table(name = "message")
@SequenceGenerator(name = "message_seq", sequenceName = "message_seq", initialValue = 7399, allocationSize = 1)
public class Message {

    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "message_seq")
    private Long id;

    @Column(columnDefinition = "text")
    private String body;

    @ManyToOne
    private User author;

    @ManyToMany
    @JoinTable(
            name = "message_recipient",
            joinColumns = {@JoinColumn(name = "USER_ID", referencedColumnName = "ID")},
            inverseJoinColumns = {@JoinColumn(name = "MESSAGE_ID", referencedColumnName = "ID")})
    private List<User> recipients;

    @Enumerated(EnumType.STRING)
    private MessageCategory category;

    @CreationTimestamp
    private Timestamp created;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getBody() {
        return body;
    }

    public void setBody(String body) {
        this.body = body;
    }

    public User getAuthor() {
        return author;
    }

    public void setAuthor(User author) {
        this.author = author;
    }

    public List<User> getRecipients() {
        return recipients;
    }

    public void setRecipients(List<User> recipients) {
        this.recipients = recipients;
    }

    public MessageCategory getCategory() {
        return category;
    }

    public void setCategory(MessageCategory category) {
        this.category = category;
    }

    public Timestamp getCreated() {
        return created;
    }

    public void setCreated(Timestamp created) {
        this.created = created;
    }
}

File: cwt-spring/src/main/java/com/cwtsite/cwt/domain/message/entity/Message.kt
package com.cwtsite.cwt.domain.message.entity

import com.cwtsite.cwt.domain.message.entity.enumeration.MessageCategory
import com.cwtsite.cwt.domain.message.service.MessageNewsType
import com.cwtsite.cwt.domain.user.repository.entity.User
import org.hibernate.annotations.CreationTimestamp
import java.time.Instant
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.EnumType
import javax.persistence.Enumerated
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.JoinTable
import javax.persistence.ManyToMany
import javax.persistence.ManyToOne
import javax.persistence.SequenceGenerator
import javax.persistence.Table

@Entity
@Table(name = "message")
@SequenceGenerator(name = "message_seq", sequenceName = "message_id_seq", allocationSize = 1)
data class Message(

    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "message_seq")
    var id: Long? = null,

    @Column(columnDefinition = "text", nullable = false)
    var body: String,

    @ManyToOne(optional = false)
    @JoinColumn(nullable = false)
    var author: User,

    @ManyToMany
    @JoinTable(
        name = "message_recipient",
        joinColumns = [JoinColumn(name = "MESSAGE_ID", referencedColumnName = "ID")],
        inverseJoinColumns = [JoinColumn(name = "USER_ID", referencedColumnName = "ID")]
    )
    var recipients: MutableList<User> = mutableListOf(),

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var category: MessageCategory = MessageCategory.SHOUTBOX,

    @Enumerated(EnumType.STRING)
    var newsType: MessageNewsType? = null,

    @field:CreationTimestamp
    @Column(nullable = false)
    var created: Instant? = null
) {
    init {
        if (recipients.isNotEmpty() && category != MessageCategory.PRIVATE) {
            throw IllegalStateException(
                "message is category $category (should be ${MessageCategory.PRIVATE})" +
                    " but has recipients $recipients"
            )
        }
        if (newsType != null && category != MessageCategory.NEWS) {
            throw IllegalStateException(
                "message has newsType $newsType but category is $category" +
                    " should be ${MessageCategory.NEWS}"
            )
        }
    }

    override fun toString() = "Message{id=$id, body=$body, category=$category, newsType=$newsType}"

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Message

        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        return id?.hashCode() ?: 0
    }
}
